vispy/visuals/graphs/layouts/force_directed.py

```diff
@@ -0,0 +1,222 @@
+        self.pos = pos
+
+    def __call__(self, adjacency_mat, directed=False):
+        """
+        Starts the calculation of the graph layout.
+
+        This is a generator, and after each iteration it yields the new
+        positions for the nodes, together with the vertices for the edges
+        and the arrows.
+
+        There are two solvers here: one specially adapted for SciPy sparse
+        matrices, and the other for larger networks.
+
+        Parameters
+        ----------
+        adjacency_mat : array
+            The graph adjacency matrix.
+        directed : bool
+            Wether the graph is directed or not. If this is True,
+            it will draw arrows for directed edges.
+
+        Yields
+        ------
+        layout : tuple
+            For each iteration of the layout calculation it yields a tuple
+            containing (node_vertices, line_vertices, arrow_vertices). These
+            vertices can be passed to the `MarkersVisual` and `ArrowVisual`.
+        """
+        if adjacency_mat.shape[0] != adjacency_mat.shape[1]:
+            raise ValueError("Adjacency matrix should be square.")
+
+        self.num_nodes = adjacency_mat.shape[0]
+
+        if issparse(adjacency_mat) and self.num_nodes < 10000:
+            # Use the sparse solver
+            solver = self._sparse_fruchterman_reingold
+        else:
+            solver = self._fruchterman_reingold
+
+        for result in solver(adjacency_mat, directed):
+            yield result
+
+    def _fruchterman_reingold(self, adjacency_mat, directed=False):
+        if self.optimal is None:
+            self.optimal = 1 / np.sqrt(self.num_nodes)
+
+        if self.pos is None:
+            # Random initial positions
+            pos = np.asarray(
+                np.random.random((self.num_nodes, self.dim)),
+                dtype='f32'
+            )
+        else:
+            pos = self.pos.astype('f32')
+
+        # Yield initial positions
+        line_vertices, arrows = straight_line_vertices(adjacency_mat, pos,
+                                                       directed)
+        yield pos, line_vertices, arrows
+
+        # The initial "temperature"  is about .1 of domain area (=1x1)
+        # this is the largest step allowed in the dynamics.
+        t = 0.1
+
+        # Simple cooling scheme.
+        # Linearly step down by dt on each iteration so last iteration is
+        # size dt.
+        dt = t / float(self.iterations+1)
+        delta = np.zeros(
+            (pos.shape[0], pos.shape[0], pos.shape[1]),
+            dtype=adjacency_mat.dtype
+        )
+
+        # The inscrutable (but fast) version
+        # This is still O(V^2)
+        # Could use multilevel methods to speed this up significantly
+        for iteration in range(self.iterations):
+            # Matrix of difference between points
+            for i in range(pos.shape[1]):
+                delta[:, :, i] = pos[:, i, None] - pos[:, i]
```

Comment:
Don't loop, use broadcasting instead:
```
delta[:, :, :] = pos[:, np.newaxis, :] - pos[:, :, np.newaxis]
```
Also `np.newaxis` is preferred to `None` even though it's currently an alias (e.g., it might not be in the future, it is more explicit, etc.)